parts/xmem_nozzle.go

```diff
@@ -3733,15 +3749,17 @@ func (xmem *XmemNozzle) handleCasPoisoning(wrappedReq *base.WrappedMCRequest, re
 		// Note: We shouldn't hit this case in practice
 		xmem.Logger().Errorf("extras.cas is not set in req %v. len of extras:%v, isSubDocOp: %v", wrappedReq.Req, len(wrappedReq.Req.Extras), isSubDocOp)
 	}
-
+	srcvb := wrappedReq.GetSourceVB()
+	tgtvb := wrappedReq.GetTargetVB()
+	seqno := wrappedReq.Seqno
 	if response.Status == mc.SUCCESS && (sentCas != 0 && sentCas != response.Cas) && !isSubDocOp { //replace mode
 		// Currently CAS regeneration takes place in two scenario's
 		// 1. If SubDoc is used
 		// 2. If there is a CAS poisoned doc and KV's protection mode is set to "replace"
 		// Note that this counter does not account for poisoned CAS's incase of subDocOp
-		return base.ReplaceMode
+		xmem.RaiseEvent(common.NewEvent(common.DocsSentWithPoisonedCas, base.ReplaceMode, xmem, []interface{}{srcvb, tgtvb, seqno}, nil))
```

Comment:
This should be `return base.ReplaceMode` itself. The `DataSent` event takes care of this automatically based on the return value of this function.

```
// record the cas poisoning protection mode by target KV if any
		protectionMode := event_otherInfo.CasPoisonProtection
		if protectionMode != base.CasNotPoisoned {
			switch protectionMode {
			case base.ErrorMode:
				metricMap[service_def.DOCS_SENT_WITH_POISONED_CAS_ERROR].(metrics.Counter).Inc(1)
				err := outNozzle_collector.handleVBEvent(event, service_def.DOCS_SENT_WITH_POISONED_CAS_ERROR)
				if err != nil {
					return err
				}
			case base.ReplaceMode:
				metricMap[service_def.DOCS_SENT_WITH_POISONED_CAS_REPLACE].(metrics.Counter).Inc(1)
				err := outNozzle_collector.handleVBEvent(event, service_def.DOCS_SENT_WITH_POISONED_CAS_REPLACE)
				if err != nil {
					return err
				}
			}
		}
```

Reply:
After fixing, run:
1. fresh and uninitialised cluster_run 
2. TestCasPoisonKVProtection in xmem_nozzle_test.go

it seems to fail now.

Reply:
Passed now.
Needed to make the following additional changes:
```
+++ b/parts/xmem_nozzle.go
@@ -3711,7 +3711,7 @@ func (xmem *XmemNozzle) receiveResponse(finch chan bool, waitGrp *sync.WaitGroup
                                if wrappedReq.OrigSrcVB != nil {
                                        additionalInfo.SetOrigSrcVB(*wrappedReq.OrigSrcVB)
                                }
-                               xmem.RaiseEvent(common.NewEvent(common.DataSent, nil, xmem, nil, additionalInfo))
+                               xmem.RaiseEvent(common.NewEvent(common.DataSent, nil, xmem, []interface{}{wrappedReq.GetSourceVB(), wrappedReq.GetTargetVB(), wrappedReq.Seqno}, additionalInfo))
                                if wrappedReq.ImportMutation && xmem.getMobileCompatible() == base.MobileCompatibilityOff && atomic.CompareAndSwapUint32(&xmem.importMutationEventRaised, 0, 1) {
```

Reply:
Did you also do this - https://github.com/couchbase/goxdcr/pull/13#discussion_r1886778137? Otherwise the same event might be double counted.